File: 290.word-pattern.java
/*
 * @lc app=leetcode id=290 lang=java
 *
 * [290] Word Pattern
 */

// @lc code=start
class Solution {
    public boolean wordPattern(String pattern, String s) {
        if (pattern.isEmpty() && s.isEmpty()) {
            return true;
        }
        if (pattern.isEmpty() || s.isEmpty()) {
            return false;
        }
        String[] words = s.split(" ");
        if (words.length != pattern.length()) {
            return false;
        }
        HashMap<Character, String> map = new HashMap<>();
        HashSet<String> usedWords = new HashSet<>();
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String word = words[i];
            if (map.containsKey(c)) {
                if (!map.get(c).equals(word)) { 
                    return false;
                }
            } else {
                if (usedWords.contains(word)) { 
                    return false;
                }
                map.put(c, word);
                usedWords.add(word);
            }
        }
        return true;
    }
}

// @lc code=end

File: 290.word-pattern.py
#
# @lc app=leetcode id=290 lang=python
#
# [290] Word Pattern
#

# @lc code=start
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        if not pattern and not s:
            return True
        if not pattern or not s:
            return False
        map = dict()
        hashset = set()
        words = s.split()
        if len(words) != len(pattern):
            return False
        for i in range(len(pattern)):
            if pattern[i] in map.keys():
                if map[pattern[i]] != words[i]:
                    return False
            else:
                if words[i] in hashset:
                    return False
                hashset.add(words[i])
                map[pattern[i]] = words[i]
        return True
        
# @lc code=end

